File: frontend/src/api/graph.ts
/**
 * 故障图谱 API
 */
import { get, post } from './index'

export interface GraphNode {
  id: string
  label: string
  type: string
  color: string
  title: string
  value: number
}

export interface GraphEdge {
  source: string
  target: string
  label: string
  weight: number
  arrows: string
}

export interface GraphData {
  nodes: GraphNode[]
  edges: GraphEdge[]
  matched_keywords?: string[]
  matched_count?: number
}

export interface GraphStats {
  total_nodes: number
  total_edges: number
  node_types: Record<string, number>
  rel_types: Record<string, number>
  density: number
}

export interface NeighborInfo {
  id: string
  label: string
  type: string
  color: string
  weight: number
}

export interface NeighborGroup {
  rel_type: string
  count: number
  neighbors: NeighborInfo[]
}

export interface NodeNeighborhood {
  node: {
    id: string
    label: string
    type: string
    color: string
    degree: number
    in_degree: number
    out_degree: number
    attrs: Record<string, any>
  } | null
  outgoing: NeighborGroup[]
  incoming: NeighborGroup[]
  summary: {
    total_out: number
    total_in: number
    total_unique_neighbors: number
    by_rel_type: Record<string, number>
  }
  error?: string
}

export async function visualizeGraph(maxNodes: number = 200): Promise<GraphData> {
  return await get<GraphData>('/graph/visualize', { params: { max_nodes: maxNodes } })
}

export async function findRelated(keywords: string[], maxHops: number = 2): Promise<GraphData> {
  return await post<GraphData>('/graph/related', { keywords, max_hops: maxHops })
}

export async function getGraphStats(): Promise<GraphStats> {
  return await get<GraphStats>('/graph/stats')
}

export async function buildGraph(): Promise<{ success: boolean; stats: GraphStats }> {
  return await post<{ success: boolean; stats: GraphStats }>('/graph/build')
}

export async function getNodeNeighborhood(nodeId: string): Promise<NodeNeighborhood> {
  return await get<NodeNeighborhood>(`/graph/node/${encodeURIComponent(nodeId)}/neighborhood`)
}

export interface GraphAnalytics {
  top_degree_centrality: Array<{
    id: string
    label: string
    type: string
    color: string
    degree: number
  }>
  connected_components: {
    count: number
    max_size: number
    sizes_distribution: Record<string, number>
  }
  node_type_density: Record<string, number>
  rel_type_density: Record<string, number>
  shortest_path_sample: {
    source: { id: string; label: string; type: string; color: string }
    target: { id: string; label: string; type: string; color: string }
    path: Array<{ id: string; label: string; type: string; color: string }>
    length: number
  } | null
}

export async function getGraphAnalytics(): Promise<GraphAnalytics> {
  return await get<GraphAnalytics>('/graph/analytics')
}

File: backend/app/api/v1/graph.py
"""
故障图谱 API（创新杀手锏）
- GET /graph/visualize - 整图（限制节点数）
- GET /graph/related - 关键词找相关子图
- GET /graph/stats - 图谱统计
- POST /graph/build - 重建图谱
"""
from typing import List, Optional
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from app.services.graph_service import get_graph_service
from app.core.logger import logger

router = APIRouter()


class RelatedRequest(BaseModel):
    keywords: List[str]
    max_hops: int = 2


@router.get("/visualize")
async def visualize(max_nodes: int = Query(200, ge=10, le=500)):
    """获取图谱可视化数据（vis.js 格式）"""
    try:
        gs = get_graph_service()
        return gs.get_full_graph(max_nodes=max_nodes)
    except Exception as e:
        logger.error(f"图谱可视化失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/related")
async def related(req: RelatedRequest):
    """根据关键词找相关子图"""
    try:
        gs = get_graph_service()
        if not req.keywords:
            raise HTTPException(status_code=400, detail="keywords 不能为空")
        return gs.find_related(req.keywords, max_hops=req.max_hops)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"图谱关联查询失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/stats")
async def stats():
    """图谱统计"""
    try:
        gs = get_graph_service()
        return gs.stats()
    except Exception as e:
        logger.error(f"图谱统计失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/build")
async def build_graph():
    """重建图谱（从案例+手册+SOP 重新构建）"""
    try:
        # 延迟导入避免循环
        from scripts.build_graph import build as run_build
        run_build()
        gs = get_graph_service()
        return {
            "success": True,
            "stats": gs.stats(),
        }
    except Exception as e:
        logger.error(f"图谱构建失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/node/{node_id}/neighborhood")
async def node_neighborhood(node_id: str):
    """获取节点详情 + 1-hop 邻居（按方向 + 关系类型分组）

    用于 Graph.vue 点击节点后的详情面板：
    - node: 节点完整属性
    - outgoing: 出边邻居（按 rel_type 分组）
    - incoming: 入边邻居（按 rel_type 分组）
    - summary: 统计摘要
    """
    try:
        gs = get_graph_service()
        return gs.get_node_neighborhood(node_id)
    except Exception as e:
        logger.error(f"节点邻域查询失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/analytics")
async def graph_analytics():
    """图谱分析指标（中心度 Top 5 / 连通分量 / 节点类型密度 / 最短路径示例）

    用于 Graph.vue 「图谱分析」面板 — 企业用户视角的洞察。
    """
    try:
        gs = get_graph_service()
        return gs.get_analytics()
    except Exception as e:
        logger.error(f"图谱分析失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))
